Source/EasyNetQ/TaskHelpers.cs

```diff
@@ -1,18 +1,82 @@
 ﻿using System;
+using System.Diagnostics;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace EasyNetQ
 {
     public static class TaskHelpers
     {
+        /// <summary>
+        /// We want to prevent callers hijacking the reader thread; this is a bit nasty, but works;
+        /// see http://stackoverflow.com/a/22588431/23354 for more information; a huge
+        /// thanks to Eli Arbel for spotting this (even though it is pure evil; it is *my kind of evil*)
+        /// </summary>
+        private static readonly Func<Task, bool> IsSyncSafe;
+
+        public static Task Completed;
+
         static TaskHelpers()
         {
-            var tcs = new TaskCompletionSource<NullStruct>();
-            tcs.SetResult(new NullStruct());
-            Completed = tcs.Task;
-        }
+            Completed = TaskEx.FromResult(new NullStruct());
```

Comment:
you can use `Completed = TaskEx.FromResult<object>(null);` to avoid an object instantiation, same in `ExecuteSynchronously` and `FromException`, then you can remove the declaration of the struct NullStruct.